### PR TITLE
Fix short-if formatting

### PR DESCRIPTION
**Pull request by csmoe**
Cloese #2777

### CODE REVIEW

**Comment by csmoe** [comment]
@nrc how can I auto-update the target files formatting when making some changes in rustfmt codebase? I used to update them with `cargo run --bin rustfmt xxx.rs", but when the formatting is controlled by config, this won't work.

File: src/expr.rs [hunk 1]
```diff
@@ -708,7 +708,7 @@ impl Rewrite for ast::Stmt {
                 };
 
                 let shape = shape.sub_width(suffix.len())?;
-                format_expr(ex, ExprType::Statement, context, shape).map(|s| s + suffix)
```
Comment (topecongiro) [inline src/expr.rs]: I think that this is incorrect. We should only put the `if` expression in a single line if it is the last statement of the block.

File: tests/target/expr.rs [hunk 2]
```diff
@@ -26,17 +26,9 @@ fn foo() -> bool {
         + aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
         + aaaaa);
 
-    {
-        for _ in 0..10 {}
-    }
+    { for _ in 0..10 {} }
```
Comment (topecongiro) [inline tests/target/expr.rs]: This should not happen, we never put a block statement in a single line.

File: tests/target/expr.rs [hunk 3]
```diff
@@ -26,17 +26,9 @@ fn foo() -> bool {
...
 
-    {
-        {
-            {
-                {}
-            }
-        }
-    }
+    { { { {} } } }
```
Comment (topecongiro) [inline tests/target/expr.rs]: So is this one.

File: tests/target/fn-single-line.rs [hunk 4]
```diff
@@ -21,9 +31,7 @@ where
 }
 
 fn fooblock() {
-    {
-        "inner-block"
-    }
+    { "inner-block" }
```
Comment (topecongiro) [inline tests/target/fn-single-line.rs]: And this one too.

**Comment by nrc** [comment]
> how can I auto-update the target files formatting when making some changes in rustfmt codebase?

There's no good way to do this, sorry. If all the tests have the same config, then you can just create a rustfmt.toml and put it in some surrounding directory, then run rustfmt. Otherwise you have to do things manually.

**Comment by csmoe** [comment]
@topecongiro ping for review.

File: src/expr.rs [hunk 5]
```diff
@@ -710,7 +710,13 @@ impl Rewrite for ast::Stmt {
...
+                let expr_type =
+                    if stmt_is_expr(&self) && context.snippet(self.span).starts_with("if ") {
```
Comment (topecongiro) [inline src/expr.rs]: This is a bit hacky - you can inspect the `ast::ExprKind` to see whether the given expression is `if` or not.

**Comment by petrochenkov** [comment]
@topecongiro @csmoe 
Any way to move this forward / prioritize?

**Comment by topecongiro** [comment]
@csmoe @petrochenkov My apologies for the late reply. I missed the notification of new commits being pushed to this PR.

Since this could not get merged before 1.0 release, the formatting change needs to be version gated. Also we need to fix the merge conflicts. I think in the current form getting these done are a bit tough, so I would rather fix the issue in a separate PR.

@csmoe Again, I am very sorry that I could not review the PR in a timely manner. I really appreciate your contribution to rustfmt.

**Pull request closed without merging**